### PR TITLE
Nurbs C-mesh foil section miniapp [nurbs-cmesh-miniapp2]

### PR DESCRIPTION
**Pull request by JacobLotz**
This PR requires PR https://github.com/mfem/mfem/pull/3226,  https://github.com/mfem/mfem/pull/3232, https://github.com/mfem/mfem/pull/3238 and https://github.com/mfem/mfem/pull/3404. **These should be merged before this PR is merged.** I have merged those branches with this branch to make it compile.

This PR contains a miniapp which creates a C-shaped mesh around a symmetric NACA foil section and generates output to visualize it in VisIt. It should be possible to replace this foil section easily with an other description of a foil section. The goal is to provide a meshing example using NURBS and showcase of PRs  https://github.com/mfem/mfem/pull/3226,  https://github.com/mfem/mfem/pull/3232, https://github.com/mfem/mfem/pull/3238 and https://github.com/mfem/mfem/pull/3404.

A possible issue in the review process is that the output can only be visualized with VisIt compiled with the current branch of MFEM. It is easier to compile glvis with this branch, to be able to visualize the mesh.

The miniapp generates a mesh with 5 patches as visualized below.
image

EDIT 31-08-2023: I think this miniapp can be really usefull. It will make benchmarking problems with flows past foil sections in MFEM a lot easier. We can perhaps also add the option to convert the mesh to a FEM mesh such that it can be used in that context as well. (This is the `SetCurvature` function.)

Possible future work:
 - Implement optimization with TMOP
 - Streamline GetTipXY() for two options

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#3419](https://github.com/mfem/mfem/pull/3419) | @JacobLotz | @pazner | @dylan-copeland + @mlstowell | 2/8/23 | 12/13/23 | 12/15/23 | |

<details>
<summary>PR Checklist</summary>
    
- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

### CODE REVIEW

**Comment by v-dobrev** [comment]
> *_New version of PR [#3407](https://github.com/mfem/mfem/pull/3226)_

Does this mean we should close #3226 and not merge it?

**Comment by v-dobrev** [comment]
> > *_New version of PR [#3407](https://github.com/mfem/mfem/pull/3226)_
> 
> Does this mean we should close #3226 and not merge it?

Actually the text and the link differ: is this PR replacing #3407, #3226, both, or none?

**Comment by JacobLotz** [comment]
> > > *_New version of PR [#3407](https://github.com/mfem/mfem/pull/3226)_
> > 
> > 
> > Does this mean we should close #3226 and not merge it?
> 
> Actually the text and the link differ: is this PR replacing #3407, #3226, both, or none?
@v-dobrev sorry for the chaos I have created. This is replacing #3407. For some reason I could not get the checks to pass in #3407 and hoped to do it in a more structured way in the current PR. Unfortunately, the checks in the current failed for some strange reason and left it to rest for the rest of the day. I see that they have just restarted.

I will close #3407 and delete the branch when done here. Sorry again for the chaos!

Kind regards,

Jacob

**Comment by pazner** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

**Comment by dylan-copeland** [comment]
As pointed out in the description, this PR depends on other PRs. In particular, there are many changes in common with PR 3404. It will be simpler to review this after the other PRs are at least fully approved, so I will wait to review.

**Comment by pazner** [comment]
> As pointed out in the description, this PR depends on other PRs. In particular, there are many changes in common with PR 3404. It will be simpler to review this after the other PRs are at least fully approved, so I will wait to review.

OK, let's remember to circle back when the other PRs are approved/merged.

**Comment by JacobLotz** [comment]
@mlstowell @dylan-copeland @pazner @IdoAkkerman  the pull requests mentioned in the PR description are all merged, so I think we should be good to go for this miniapp as well. I have added something small to the PR description and have done a bit of cleaning up on the miniapp itself.

**Comment by pazner** [comment]
> @mlstowell @dylan-copeland @pazner @IdoAkkerman the pull requests mentioned in the PR description are all merged, so I think we should be good to go for this miniapp as well. I have added something small to the PR description and have done a bit of cleaning up on the miniapp itself. The continuous integration check is failing, and I do not understand why. I think it is just canceled and would probably pass upon restarting the check?

Thanks for the update @JacobLotz. The CI is now passing.

This looks like it's ready for review @dylan-copeland, @mlstowell.

File: miniapps/nurbs/nurbs_naca_cmesh.cpp
Comment (dylan-copeland) [inline miniapps/nurbs/nurbs_naca_cmesh.cpp:63]: The compiler says `const` is meaningless here in the return value.

File: .gitignore
Comment (dylan-copeland) [inline .gitignore:280]: `Naca_cmesh` is not in the repo, so this line can be deleted.

File: miniapps/nurbs/nurbs_naca_cmesh.cpp
Comment (dylan-copeland) [inline miniapps/nurbs/nurbs_naca_cmesh.cpp:20]: I'm not sure whether MFEM has a policy on correct vs. British spellings, e.g. `visualization` vs. `visualisation`. I suppose we can allow the British spelling, but it would be more consistent with the rest of the library to use `visualization`.

File: miniapps/nurbs/nurbs_naca_cmesh.cpp
Comment (dylan-copeland) [inline miniapps/nurbs/nurbs_naca_cmesh.cpp:3]: `-nrad` is not an option.

File: miniapps/nurbs/nurbs_naca_cmesh.cpp
Comment (dylan-copeland) [inline miniapps/nurbs/nurbs_naca_cmesh.cpp:640]: `halve` -> `half` throughout this file.

File: miniapps/nurbs/nurbs_naca_cmesh.cpp
Comment (dylan-copeland) [inline miniapps/nurbs/nurbs_naca_cmesh.cpp:22]: Could you please provide more detail on this in VisIt?

File: miniapps/nurbs/nurbs_naca_cmesh.cpp
Comment (dylan-copeland) [inline miniapps/nurbs/nurbs_naca_cmesh.cpp:35]: describes

File: miniapps/nurbs/nurbs_naca_cmesh.cpp
Comment (dylan-copeland) [inline miniapps/nurbs/nurbs_naca_cmesh.cpp:101]: These empty comment lines are not usually used in the MFEM examples.

File: miniapps/nurbs/nurbs_naca_cmesh.cpp
Comment (dylan-copeland) [inline miniapps/nurbs/nurbs_naca_cmesh.cpp:595]: If you split these lines between terms, `astyle` will not format so awkwardly.

File: miniapps/nurbs/nurbs_naca_cmesh.cpp
Comment (dylan-copeland) [inline miniapps/nurbs/nurbs_naca_cmesh.cpp:568]: `mesh->Print(os);` for an `ofstream os` will write a file identical to the one in `Naca_cmesh/mesh_000000`, which can be visualized with glvis. The format of that file is different from `naca-cmesh.mesh`, but I'm not sure if there is a real difference in the loaded meshes.

File: miniapps/nurbs/nurbs_naca_cmesh.cpp
Comment (JacobLotz) [reply miniapps/nurbs/nurbs_naca_cmesh.cpp:3]: Updated

File: miniapps/nurbs/nurbs_naca_cmesh.cpp
Comment (JacobLotz) [reply miniapps/nurbs/nurbs_naca_cmesh.cpp:640]: Updated

File: miniapps/nurbs/nurbs_naca_cmesh.cpp
Comment (JacobLotz) [reply miniapps/nurbs/nurbs_naca_cmesh.cpp:35]: Updated

File: miniapps/nurbs/nurbs_naca_cmesh.cpp
Comment (JacobLotz) [reply miniapps/nurbs/nurbs_naca_cmesh.cpp]: I ran a spell checker with American English.

File: miniapps/nurbs/nurbs_naca_cmesh.cpp
Comment (JacobLotz) [reply miniapps/nurbs/nurbs_naca_cmesh.cpp]: Removed

File: miniapps/nurbs/nurbs_naca_cmesh.cpp
Comment (JacobLotz) [reply miniapps/nurbs/nurbs_naca_cmesh.cpp:22]: More info on multiress-control is added.

**Comment by JacobLotz** [comment]
@dylan-copeland I have just tested the dedicated output for glvis and the mesh the miniapp produces in glvis. Tt can open both, so it might be a bit redundant. However, I think we should keep the dedicated glvis output for clarity as it think it is better than having the dummy input option for allowing glvis visualisation.

**Comment by pazner** [comment]
@dylan-copeland, @mlstowell, can you take another look at this?

@JacobLotz, can you merge master and fix the conflicts?

Thanks!

**Comment by dylan-copeland** [comment]
@JacobLotz sorry for the delay. This is the second time in my github career that my comments were "pending" in a review. I thought you could see them, because I could, but they were not actually posted.

File: miniapps/nurbs/nurbs_naca_cmesh.cpp
Comment (JacobLotz) [reply miniapps/nurbs/nurbs_naca_cmesh.cpp]: Fixed!

File: miniapps/nurbs/nurbs_naca_cmesh.cpp
Comment (JacobLotz) [reply miniapps/nurbs/nurbs_naca_cmesh.cpp]: I have explained it a bit more. You can find this option under Operators/Selection/MultiresControl. It allows to visualize the NURBS curve more precisely by evaluating it at more points along the line. Typo is fixed.

**Review by mlstowell — APPROVED** [review]
My only concern is that the VisIt output is not usable without recompiling VisIt. It may be best to add a comment mentioning this so that users don't think something is broken. Unfortunately I don't think we have a specific MFEM version number to reference yet but perhaps we can come up with something...

File: miniapps/nurbs/nurbs_naca_cmesh.cpp
Comment (dylan-copeland) [inline miniapps/nurbs/nurbs_naca_cmesh.cpp:73]:
```suggestion
void GetTipXY(const NACA4 & foil_section, KnotVector *kv, double tf,
```

File: miniapps/nurbs/nurbs_naca_cmesh.cpp
Comment (dylan-copeland) [inline miniapps/nurbs/nurbs_naca_cmesh.cpp:187]:
```suggestion
   constexpr double deg2rad = M_PI/180;
```

File: miniapps/nurbs/nurbs_naca_cmesh.cpp
Comment (dylan-copeland) [inline miniapps/nurbs/nurbs_naca_cmesh.cpp:184]:
```suggestion
   const double tail_fraction = 1.0 - tip_fraction;
```

File: miniapps/nurbs/nurbs_naca_cmesh.cpp
Comment (dylan-copeland) [inline miniapps/nurbs/nurbs_naca_cmesh.cpp:273]: A few lines have `;;`

File: miniapps/nurbs/nurbs_naca_cmesh.cpp
Comment (dylan-copeland) [inline miniapps/nurbs/nurbs_naca_cmesh.cpp:578]: It looks like the files glvis_naca-cmesh.mesh and naca-cmesh.mesh are the same, except for comments. Do both need to be output?

File: miniapps/nurbs/nurbs_naca_cmesh.cpp
Comment (dylan-copeland) [inline miniapps/nurbs/nurbs_naca_cmesh.cpp:535]: This is a lot of `delete` calls. A simpler implementation might be to use `unique_ptr`. For example, see miniapps/solvers/div_free_solver.hpp.

File: miniapps/nurbs/nurbs_naca_cmesh.cpp
Comment (dylan-copeland) [inline miniapps/nurbs/nurbs_naca_cmesh.cpp:656]:
```suggestion
   const int n = ncp/2;
```

**Comment by v-dobrev** [comment]
> My only concern is that the VisIt output is not usable without recompiling VisIt. It may be best to add a comment mentioning this so that users don't think something is broken. Unfortunately I don't think we have a specific MFEM version number to reference yet but perhaps we can come up with something...

@JacobLotz, can you clarify: does one really need VisIt compiled with newer version of MFEM or using the pre-build binaries from the VisIt website should be fine? If newer version of MFEM is needed, is v4.6 sufficient or does one need this PR?

File: miniapps/nurbs/nurbs_naca_cmesh.cpp
Comment (JacobLotz) [reply miniapps/nurbs/nurbs_naca_cmesh.cpp:24]: Fixed.

File: miniapps/nurbs/nurbs_naca_cmesh.cpp
Comment (JacobLotz) [reply miniapps/nurbs/nurbs_naca_cmesh.cpp:22]: Fixed

File: miniapps/nurbs/nurbs_naca_cmesh.cpp
Comment (JacobLotz) [reply miniapps/nurbs/nurbs_naca_cmesh.cpp:55]: Fixed

File: miniapps/nurbs/nurbs_naca_cmesh.cpp
Comment (JacobLotz) [reply miniapps/nurbs/nurbs_naca_cmesh.cpp:273]: Fixed!

File: miniapps/nurbs/nurbs_naca_cmesh.cpp
Comment (dylan-copeland) [inline miniapps/nurbs/nurbs_naca_cmesh.cpp:648]: Unused.

File: miniapps/nurbs/nurbs_naca_cmesh.cpp
Comment (JacobLotz) [reply miniapps/nurbs/nurbs_naca_cmesh.cpp:578]: I did this on purpose. In the script the mesh file is being generated and tested, otherwise the script would not really make sense. However, the option `--visualisation` for GlVis visualisation is mandatory to have. I thought it would make more sense to have an action associated with with this that actually gives a glvis visualisation file, instead of leaving it a dummy option which would not do anything.

File: miniapps/nurbs/nurbs_naca_cmesh.cpp
Comment (JacobLotz) [reply miniapps/nurbs/nurbs_naca_cmesh.cpp:648]: This was still work in progress after I was summoned for diner ;)

File: miniapps/nurbs/nurbs_naca_cmesh.cpp
Comment (JacobLotz) [reply miniapps/nurbs/nurbs_naca_cmesh.cpp:6]: Fixed

**Comment by JacobLotz** [comment]
> > My only concern is that the VisIt output is not usable without recompiling VisIt. It may be best to add a comment mentioning this so that users don't think something is broken. Unfortunately I don't think we have a specific MFEM version number to reference yet but perhaps we can come up with something...
> 
> @JacobLotz, can you clarify: does one really need VisIt compiled with newer version of MFEM or using the pre-build binaries from the VisIt website should be fine? If newer version of MFEM is needed, is v4.6 sufficient or does one need this PR?

For visualizing this MFEM 4.6 is sufficient. I have asked the VisIt team on their plans of updating their default version of MFEM to 4.6. They where very kind and plan to update the dependency soon, see [their pr 19122](https://github.com/visit-dav/visit/pull/19122).
I have left a note in the description of the mininapp stating the requirement.

File: miniapps/nurbs/nurbs_naca_cmesh.cpp
Comment (JacobLotz) [reply miniapps/nurbs/nurbs_naca_cmesh.cpp:548]: Interesting suggestion! I have looked into it. I think that `std::unique_ptr` is not compatible with some of the constructors of `NURBSPatch` such as 

```
NURBSPatch(const KnotVector *kv0, const KnotVector *kv1, int dim_);
```

Therefor I think we will get a mix of 'normal' pointers and `std::unique_ptr` which I think is undesirable. I will leave it in its current state.

Comment (dylan-copeland) [reply miniapps/nurbs/nurbs_naca_cmesh.cpp:548]: I am fine with the current state and do not think it is necessary to change to `unique_ptr` for this PR. A normal, raw pointer can be obtained by calling `.get()`: https://en.cppreference.com/w/cpp/memory/unique_ptr/get

File: miniapps/nurbs/nurbs_naca_cmesh.cpp
Comment (dylan-copeland) [reply miniapps/nurbs/nurbs_naca_cmesh.cpp:591]: I see, it is necessary to write the file and read it back for constructing the mesh. This is generally necessary for NURBS meshes, because there is no constructor in the code. Writing a file is the only to construct a NURBS mesh. In the future, we should add NURBS mesh constructors to MFEM.

**Review by dylan-copeland — APPROVED** [review]

File: miniapps/nurbs/nurbs_naca_cmesh.cpp
Comment (pazner) [reply miniapps/nurbs/nurbs_naca_cmesh.cpp:548]: I think moving to `unique_ptr` will be better. @JacobLotz, can you see if you can make it work with `unique_ptr::get` as Dylan suggested?

Comment (JacobLotz) [reply miniapps/nurbs/nurbs_naca_cmesh.cpp]: @dylan-copeland @pazner with `.get`, the implementation actually was quite easy. Thanks for this suggestion!

File: miniapps/nurbs/nurbs_naca_cmesh.cpp
Comment (JacobLotz) [reply miniapps/nurbs/nurbs_naca_cmesh.cpp:585]: I agree!

**Comment by pazner** [comment]
Merged in `next` for testing...

File: miniapps/nurbs/nurbs_naca_cmesh.cpp
Comment (v-dobrev) [reply miniapps/nurbs/nurbs_naca_cmesh.cpp:585]: Technically, you don't **need** to write to a file -- you can write to a `std::stringstream` and then read from it -- all of this will be in-memory. 😁

File: miniapps/nurbs/nurbs_naca_cmesh.cpp
Comment (v-dobrev) [inline miniapps/nurbs/nurbs_naca_cmesh.cpp]: For clarity, one can make the parameter `const unique_ptr` -- this way one can still call non-const methods of `KnotVector` but one cannot extract/move out the pointer from the object referenced by `kv`. Without the `const` the method can "steal" the pointer from `kv`.

File: miniapps/nurbs/nurbs_naca_cmesh.cpp
Comment (dylan-copeland) [reply miniapps/nurbs/nurbs_naca_cmesh.cpp:585]: Good point @v-dobrev, and I have done this for small meshes, e.g. a square or cube of a certain order. Even for such a simple mesh, such as a cube of order 3, it is tricky to write out all the DOFs in the right order. I plan to make some constructors to make the process more user-friendly, some day.

File: miniapps/nurbs/nurbs_naca_cmesh.cpp
Comment (v-dobrev) [reply miniapps/nurbs/nurbs_naca_cmesh.cpp]: Actually, if there is no transfer of ownership of the `KnotVector` pointed to by `kv`, it is much cleaner to just use parameter `KnotVector &` (or `const KnotVector &`, if the object is not being modified) instead of (`const`) `unique_ptr<KnotVector> &`.

Comment (v-dobrev) [reply miniapps/nurbs/nurbs_naca_cmesh.cpp]: And finally, I see that I've been commenting on outdated code and the latest version is actually using `const KnotVector &`. 🤣

File: miniapps/nurbs/nurbs_naca_cmesh.cpp
Comment (JacobLotz) [reply miniapps/nurbs/nurbs_naca_cmesh.cpp:585]: @v-dobrev and @dylan-copeland These comments got a bit snowed under in others, so I notice them to late. @v-dobrev his suggestion is indeed possible. However, using this script without the option for glvis visualisation will then also remove the desired output of this script: the mesh itself. This might lead to unclarity in for user of this mesher as the expected output cannot be found. The script just does not do anything.

**Pull request merged**
